### ISSUE TITLE
Core database model

### ISSUE THREAD

**Issue by lparsons** [Princeton-LSI-ResearchComputing/tracebase#16]
This is a placeholder issue to gather information about the core database model.

Measurement object in particular will need to have functions that lookup and/or compute values including:
* `serum_infusate_abundance` - the 'normalized_abundance' for the infusate compound (tracer) in the serum sample from this mouse.  If 13C-lactate was infused this is the value for 'normalized_abundance' of serum lactate.

![QuickDBD-TraceBase (9)](https://user-images.githubusercontent.com/645128/117880955-ac356e80-b276-11eb-8e86-874a60bdc09e.png)

**Comment by jcmatese** [comment]
If you are trying to fully model the isomer edge cases (citrate/isocitrate-containing peaks), the the compound-to-peak relationship probably needs to be many-to-many.

**Comment by lparsons** [comment]
> If you are trying to fully model the isomer edge cases (citrate/isocitrate-containing peaks), the the compound-to-peak relationship probably needs to be many-to-many.

That's a good point @jcmatese. The question I have is, how often does the lab make a many-to-many association and how much additional complication would that create. My gut tells me that they do it not too often, but regularly, and that the complication wouldn't be too bad, but that it also something we could hold off on for now and implement later once we have the bones in place. But it's good to keep in mind.

**Comment by hepcat72** [comment]
My instincts tell me that while the lab may not be effectively thinking about the M:M compound:peak relationship in their day-to-day work, I would not be surprised that they may discover that they need it... when it's important.  I'm inclined to add a linking table between the 2.  But a compromise might be to make the primary key of the peak table be the combined values of run_id, sample_id, and the two median values that locate the position of the peak.  In fact, the measurement assumes a compound.  Perhaps the compound link should be in there.

**Comment by lparsons** [comment]
I certainly don't disagree, and I'm glad we're having the discussion since these are the types of things it's good to think about before forging ahead. I feel that adding the many-to-many relationship would allow us to make the model more "accurate", but I'm not sure it would make it more useful to the researchers, at least at the moment. The compounds they care about are of a limited set, and for now, I think they understand that something labeled citrate means it could also be isocitrate and having an extra HMDB id associated isn't likely to provide much additional utility. I'm inclined to keep it simple for now and tackle the issue if/when it actually arises. For right now, with the data we have, we can make this work, imho.

**Comment by hepcat72** [comment]
It certainly is a simple expectation right now, to tackle if/when the need arises, given the context of the current state and plans of the repo, but I'm convinced in principle that a little effort now is likely to decrease time & effort costs downstream.

The more I think about it, the more I think that simply moving compound_id to the measurement table may be a good idea.  Although, I'm not certain I fully understand the data the measurement table contains.  Can someone ELI5 it for me?  It's clearly a measurement associated with a peak.  Are the abundances and label info associated exclusively with a single compound?

**Comment by lparsons** [comment]
> It certainly is a simple expectation right now, to tackle if/when the need arises, given the context of the current state and plans of the repo, but I'm convinced in principle that a little effort now is likely to decrease time & effort costs downstream.

I don't disagree with that. It would be nice to get this right. @mneinast or @cbartman1 - any comments on this? I'll also ping Matt on Slack.

> The more I think about it, the more I think that simply moving compound_id to the measurement table may be a good idea. Although, I'm not certain I fully understand the data the measurement table contains. Can someone ELI5 it for me? It's clearly a measurement associated with a peak. Are the abundances and label info associated exclusively with a single compound?

The measurement table is certainly an interesting one, and I'm not sure that this structure is quite correct. I'll try to explain it and perhaps that will help me to wrap my head around it all more clearly too. A peak is found a given m/z (mass) and rt (retention time) which provides strong hints as to which compound it represents. The researcher then associates this peak with a Compound in El- Maven (which is really just a formula, NOT a specific HMDB ID). Based on that formula, El-Maven looks for other peaks at the same rt, but with m/z values that differ by the amount expected if they had one or more labeled atoms. For example, isotopically labeled carbon (C13) has a mass of one neutron greater than C12. More labeled carbon atoms mean higher m/z. The measurement (which might not be a good name) attempts to capture the detected amount of a given compound (formular really) based on the0` labeled_atom` and `labeled_count`.

**Comment by fkang-pu** [comment]
I would rather not to have compound_id in measurement table.  Instead, prefer other options: 1) add a linking table between peak and compound; or 2) replace compound_id with formula in peak table, references compound.formula.

I don't know the details, but generally a lab runs calibrations with knows compounds for a LC-MS (or LC-MS/MS) method. So they know they can use a unique mz/rt for a defined method (we point it to a protocol) to identify a compound or a group of compounds. I think as long as we link a unique set of mz/rt/protocol to a peak_id (or detection_id), we are fine. Then we can link a peak_id to one or more compounds (e.g. citrate/isocitrate) and a row of measurement data from a MS run.

 I also feel it's better to move sample_id from Peak table to MSRun table (one sample can have many MS runs, but one unique MSRun has only one sample, and one protocol).  With ms_run_id in Peak table, seems that peak_id should be unique for a row in Measurement table. Another question: could move ms_run_id from Peak table to Measurement table?  The argument is that many MSRuns using the same method may share some identified peaks . If keeping ms_run_id in Measurement table were reasonable, peak_id + ms_run_id would be unique for a row in Measurement table.

**Comment by lparsons** [comment]
> I would rather not to have compound_id in measurement table. Instead, prefer other options: 1) add a linking table between peak and compound; or 2) replace compound_id with formula in peak table, references compound.formula.

Given the [feedback from](https://tracebase.slack.com/archives/C01M97YLQE5/p1616173025014900?thread_ts=1616170778.014600&cid=C01M97YLQE5) @mneinast, I don't think a linking table is required. I like the idea of using formula, but I think that would complicate programming. Perhaps, instead, we should consider allowing multiple HMDB_IDs be associated with one compound record. I'm also going to rename Peak to "Observed Compound" which is more like your original suggestion @fkang-pu. ;-) I'm also tempted to change Measurement to Peak, but I'll hold off for now.

> I don't know the details, but generally a lab runs calibrations with knows compounds for a LC-MS (or LC-MS/MS) method. So they know they can use a unique mz/rt for a defined method (we point it to a protocol) to identify a compound or a group of compounds.

Yes, I believe that is how it's being done in the Rabinowitz lab for these experiments.

> I think as long as we link a unique set of mz/rt/protocol to a peak_id (or detection_id), we are fine. Then we can link a peak_id to one or more compounds (e.g. citrate/isocitrate) and a row of measurement data from a MS run.

Yes, that makes sense to me.

> I also feel it's better to move sample_id from Peak table to MSRun table (one sample can have many MS runs, but one unique MSRun has only one sample, and one protocol). With ms_run_id in Peak table, seems that peak_id should be unique for a row in Measurement table.

Excellent suggestion, I'll make that change.

> Another question: could move ms_run_id from Peak table to Measurement table? The argument is that many MSRuns using the same method may share some identified peaks . If keeping ms_run_id in Measurement table were reasonable, peak_id + ms_run_id would be unique for a row in Measurement table.

I'm not quite sure I've caught up to you here. In fact, my "Peak" table is misnamed, and I'm changing it to "Observed Compound". It actually represents a group of peaks (measurements) that represent the various labeled forms of the same compound. That said, I may just not have grasped this last point yet.

**Comment by lparsons** [comment]
I've updated the proposed model.

**Comment by hepcat72** [comment]
I'm definitely not opposed to the compound/peak linking table.  I still would like to understand the measurement table better.  I'm just trying to get caught up here, so let me see if I can break things down conceptually by making some naive statements and see if I have a rudimentary understanding to start out.

# The easy(?) parts:
1. Compound is a chemical structure that exists in nature and is the pure thing we're trying to identify in the mass spec of the sample
2. Peak is a point on a (2D?) trace, the dimensions of which are m/z (mass/charge) and rt (retention time).  (From my small bit of mass spec experience, a spot is the result of particles shot through a charged field and where they land was influenced by mass, charge, and gravity - is that dated knowledge correct?  And those particles may or may not be pieces of a larger compound that have been broken apart (my Los Alamos experience was with sequencing proteins via mass spec that did this).).  The peak table will only describe a peak using one type of peak determination (e.g. summit versus centroid) - not multiple.
# The part I'm not sure of is the measurement.  Correct me if I'm wrong on any of these statements.
1. The sample is exposed to a certain amount of labeled atoms (e.g. C13).  The sample (e.g. mouse cells) incorporates those labeled atoms and the resulting mass spec shows differences that help elucidate the compounds present - and what is taking up the labeled atoms.
2. The "measurement" says "let's assume the compound(s) present is(/are) X.  Given that assumption and the known amount of label used, let's estimate compound X's abundance."

**Comment by hepcat72** [comment]
Just found [this article](https://tarabytesomics.wordpress.com/2019/06/07/understanding-metabolomics-data-an-overview-for-bioinformaticians/) which I'm about to read.  Is this the sort of mass spec data we're working with?:

![image](https://user-images.githubusercontent.com/2300532/111834457-02ff7700-88ca-11eb-8a7c-ab93cc9e9222.png)

**Comment by jcmatese** [comment]
yes
